Add tests and update csg boltzmann

JoshuaSBrown opened the pull request:
In an attempt to remove use of histogram and replace it with histogramnew without breaking the code it was necessary to setup a few unit tests for classes that made use of the histogram class. This merge is responsible for creating these tests which are located in the csg-boltzmann folder. 
In response to issue https://github.com/votca/tools/issues/47

Commit:
1. Added comments
2. Removed magic numbers such as the boltzmann constant and gave it a
variable name
3. Remove redundant use of temperature in internal method. The boltzmann
invert method was taking one of the internal variables as an argument
when it did not need to as it was already stored in the class.

Commit:
Added working test for bonded statistics testing method BeginCG

Commit:
Added unit test for EvalConfiguration method

Commit:
Added to test of tabulated potential

Commit:
Added comment explaining what bonded statistics class is used for

Commit:
Appended private methods with underscore to differentiate

Commit:
Corrected underscore labeling, two public methods were accidentally changed

Commit:
Added more underscores that were missing

Commit:
Comments explaining how the class is used have been added

Commit:
Added more tests

Commit:
Added csg-boltzmann as library

Commit:
Removed commented out code

Commit:
Added in brackets to prevent errors with if else statements and for loops

Commit:
Upated header guards

Commit:
Updated bondedstatistics header guard

Commit:
Added tabulatedpotential test to CMake file

Commit:
Added test for csg topology file

Commit:
Applied google style formatting to tabulatedpotential.h file

Commit:
Applied google style formatting to tabulatedpotential.cc

Commit:
Applied google style formatting to test_tabulatedpotential.cc

Commit:
Removed using namespace from tabulatedpotential and bondedstatistics header files

Commit:
Cleaned up namespace of bondstatistics and tabulatedpotential class files

Commit:
Updated date of license

JoshuaSBrown committed:
Merge branch 'master' into add-tests-and-update-csg-boltzmann

junghans commented:
Does that fix #155?

junghans commented:
And could we add a regression test for csg_boltzmann? I.e echoing some commands into the command and compare the output.

JoshuaSBrown commented:
The boltzmann constant used in the boltzmann inversion function in the tabulated potential has a hard coded value in what I believe is units of [ kJ/(mol K) ]. Is this what you are referring too?

Commit:
Corrected typo in bead_ptr3 variable name

Commit:
Removed unused variable number_of_interacctions from test_tabulatedpotential.cc

Commit:
Fixed typo in bead_ptr3 setting id was instead using bead_ptr2

Commit:
fixed interaction group and made each group unique

JoshuaSBrown commented:
It is not really that clear to me how to use csg_boltzmann. @junghans in the interactive mode I can specify a histogram etc... but is there a way to do this without going into the interactive mode? I'm also not sure what to use for the selection to get meaningful output? I've been trying with files in tools/references/spce but I don't know if this is appropriate either. 

    list: list all available bonds
    vals <file> <selection>: write values to file
    hist <file> <selection>: create histogram
    tab <file> <selection>: create tabulated potential
    autocor <file> <selection>: calculate autocorrelation, only one row allowed in selection!
    cor <file> <selection>: calculate correlations, first row is correlated with all other rows

junghans commented:
I was thinking about something like `echo "command1\ncommand2\nq" | csg_boltzmann`.

junghans commented:
I have did something like:
```
hist end.txt end-end
vals end-v.txt end-end
hist AA.txt AA-bond
hist AAA.txt AAA-angle
hist set periodic 1
hist set auto 0
hist set min -3.14159
hist set max 3.14159
hist AAAA.txt AAAA-dihedral
tab set T 400
tab U_AA.txt AA-bond
tab U_AAA.txt AAA-angle
tab set periodic 1
tab set smooth1 1
tab set auto 0
tab set min -3.14159
tab set max 3.14159
tab U_AAAA.txt AAAA-dihedral
autocor end-cor.txt end-end
autocor AA-cor.txt AA-bond[4]
autocor AAA-cor.txt AAA-angle[5]
autocor AAAA-cor.txt AAAA-dihedral[4]
cor bond-angle.txt AAA-angle[4] AA-bond
cor bond-dihe.txt AAAA-dihedral[4] AA-bond
cor angle-dihe.txt AAAA-dihedral[4] AAA-angle
q
```
and then `cat "cmd" | csg_boltzmann --top topol.tpr` before.

However it seems CMake has a hard time to redirect stdin into tests (see https://stackoverflow.com/questions/36304289/how-to-use-redirection-in-cmake-add-test).

So @JoshuaSBrown how hard would it be to add an option (e.g. `--in` to have `csg_boltzmann` read from a file instead of stdin?

JoshuaSBrown commented:
I can do that but let's put it in a separate merge, that will be more than a quick fix, this merge is good to go unless you see something else that I have done that should be changed?

junghans commented:
If you do that than I will write the tests!

junghans commented on src/csg_boltzmann/CMakeLists.txt:
```diff
@@ -1,4 +1,5 @@
 file(GLOB CSG_BO_SOURCES *.cc)
+add_library(votca_csg_boltzmann ${CSG_BO_SOURCES})
```
Won't that library have a `main()` if we use all `csg_boltzmann` sources?

JoshuaSBrown replied:
Yes, I guess I could remove the main.cc that would be less sloppy.

junghans commented:
Will this help with https://github.com/votca/tools/issues/47?

junghans commented:
Maybe we can add an option to change the unit of `kB` down the road?

JoshuaSBrown commented:
Yes, that could be a separate issue. How units are handled could be streamlined across the whole VOTCA package. Though I don't yet have a good feel for that aspect of the VOTCA package yet as I have not really used the tools with the various file formats we support.

JensWehner commented on src/csg_boltzmann/tabulatedpotential.cc:
```diff
@@ -15,333 +15,348 @@
+using namespace votca::csg;
+
+/* Boltzmann constant in units [ kJ/(mol K) ] */
+const double kB = 0.0083109;
```
move to votca/tools/constants

JoshuaSBrown replied:
Sounds good.

JensWehner commented on src/csg_boltzmann/tabulatedpotential.cc:
```diff
@@ -15,333 +15,348 @@
+    if (data[i] > 0)
+      _min = min(data[i], _min);
+  }
+  _max = -kB * _Temperature * log(_max);
```
really good catch, well done

JoshuaSBrown replied:
Well you know I try... :)

JensWehner reviewed (approved):
lgtm nice rework.  So we can remove histogram and rename histogram_new to histogram?

JoshuaSBrown commented:
@JensWehner and yes that is the idea, once we have good tests in place we should be able to replace histogram with histogramnew and then remove it hopefully without any hiccups.

JoshuaSBrown committed:
1. Used constants in tools constants header file 2. removed main.cc from library in CmakeLists.txt file

JoshuaSBrown committed:
1. Updated header gaurds
2. Placed all code in csg-boltzmann within votca csg namespace
3. Cleanup header files in terms of removing using namespace calls

JoshuaSBrown committed:
Merge pull request #292 from votca/cleanup-namespace

1. Updated header gaurds

JoshuaSBrown commented:
https://github.com/votca/tools/pull/84 should be merged first.
